mopidy_soundcloud/soundcloud.py

```diff
@@ -348,14 +369,76 @@ def parse_track(self, data, remote_url=False):
+        regex_str = r"client_id=([a-zA-Z0-9]+)"
+        soundcloud_soup = BeautifulSoup(public_page, "html.parser")
+        scripts = soundcloud_soup.find_all("script", attrs={"src": True})
+        client_id = None
+        for script in scripts:
+            try:
+                regex_iter = re.finditer(regex_str, get_page(script["src"]))
+                client_id = next(regex_iter)[0].split("=")[-1]
+            except StopIteration:
+                pass
+
+        self.public_client_id = client_id
+        logger.debug(
+            f"Updated SoundCloud public client id to: {self.public_client_id}"
+        )
+
+    def _get_public_stream(self, progr_stream):
+        params = [("client_id", self.public_client_id)]
+        return self.public_stream_client.get(progr_stream, params=params)
+
     @cache()
-    def get_streamble_url(self, url):
-        req = self.http_client.head(streamble_url(url, self.CLIENT_ID))
+    def get_streamable_url(self, sharing, permalink_url, stream_url):
+        progressive_urls = {"preview": "", "stream": ""}
+        if sharing == "public":
+            res = self.public_stream_client.get(permalink_url)
+
+            for html_substring in res.text.split('"'):
+                if html_substring.endswith("/progressive"):
+                    if "preview" in html_substring:
+                        progressive_urls["preview"] = html_substring
+                    elif "stream" in html_substring:
+                        progressive_urls["stream"] = html_substring
+
+            if progressive_urls["stream"]:
+                stream = self._get_public_stream(progressive_urls["stream"])
+                if stream.status_code in [401, 403, 429]:
+                    self._update_public_client_id()  # refresh public client id once
+                    stream = self._get_public_stream(progressive_urls["stream"])
+
+                try:
+                    return stream.json()["url"]
+                except Exception as e:
+                    logger.info(
+                        "Streaming of public song using public client id failed, "
+                        "trying with standard app client id.."
+                    )
+                    logger.debug(
+                        f"Caught public client id stream fail:\n{e}"
+                        f"\n{parse_fail_reason(stream.reason)}"
+                    )
+
+        # ~quickly yields rate limit errors
+        req = self.http_client.head(streamble_url(stream_url, self.CLIENT_ID))
         if req.status_code == 302:
             return req.headers.get("Location", None)
         elif req.status_code == 429:
-            reason = "" if req.reason == "Unknown" else f" ({req.reason})"
-            logger.warning(f"SoundCloud daily rate limit exceeded{reason}")
+            logger.warning(
+                f"SoundCloud daily rate limit exceeded {parse_fail_reason(req.reason)}"
+            )
+            if progressive_urls["preview"]:
```

Comment:
If preview stream is found, reverts to this stream on fail with standard client id.